src/methods/join.js

```diff
@@ -806,7 +806,40 @@ export class JoinImplementation {
     c1: CreatedObjects,
     c2: CreatedObjects
   ): PropertyBindings {
+    let joinLeakedObjectProperty = (b: PropertyBinding, d1: void | Descriptor, d2: void | Descriptor) => {
+      if (d1 === undefined) {
+        if (b.object instanceof ObjectValue && c2.has(b.object)) return d2; // no join
+
+        invariant(m2.has(b)); // The alternative path must be the source of leaking
+        return d2;
+      } else if (d2 === undefined) {
+        if (b.object instanceof ObjectValue && c1.has(b.object)) return d1; // no join
+
+        invariant(m1.has(b)); // The consequent path must be the source of leaking
+        return d1;
+      } else {
+        // Make the leak unconditional
+        let isTrue = bv => bv instanceof BooleanValue && bv.equals(realm.intrinsics.true);
+        if (isTrue(d1.value)) {
+          return d1;
+        } else if (isTrue(d2.value)) {
+          return d2;
+        } else {
+          invariant(
+            b.object instanceof ObjectValue && c1.has(b.object) && c2.has(b.object),
+            "Evidence of object leaked, but _isLeaked is false"
+          );
+          return d1;
+        }
+      }
+    };
+
     let join = (b: PropertyBinding, d1: void | Descriptor, d2: void | Descriptor) => {
+      if (b.object instanceof ObjectValue && b.key === "_isLeaked") {
```

Comment:
Hm, so in `PropertyBinding`, the `object` is defined to be `AbstractObjectValue`, and thus you need to do the cast here. Okay. But I wonder if that's really the most appropriate type definition.
For fun, I removed the `| AbstractObjectValue`, and ran flow again --- no violation!
Now, of course that doesn't have to mean much, but I wonder if we can make this new code cleaner by if can simply "fix" the type definition, after manually reviewing all assignments?

Reply:
See #2208